File: BackEnd/src/domain/usecases/notifications/MarkNotificationsReadUseCase.ts
/**
 * @file MarkNotificationsReadUseCase.ts
 * @description Caso de uso para marcar todas las notificaciones del usuario como leídas.
 * @module Reminders
 * @layer Domain
 *
 * @implements {IMarkNotificationsReadUseCase}
 * @injectable
 * @dependencies INotificationRepository
 */

import { injectable, inject } from 'inversify';
import { TYPES } from '../../../core/types.js';
import type { IMarkNotificationsReadUseCase } from '../../interfaces/usecases/notifications/IMarkNotificationsReadUseCase.js';
import type { INotificationRepository } from '../../repositories/INotificationRepository.js';

/**
 * Marca todas las notificaciones del usuario como leídas (isRead=true).
 *
 * @implements {IMarkNotificationsReadUseCase}
 * @injectable
 * @dependencies INotificationRepository
 */
@injectable()
export class MarkNotificationsReadUseCase implements IMarkNotificationsReadUseCase {
  constructor(
    @inject(TYPES.INotificationRepository) private readonly notifRepo: INotificationRepository,
  ) {}

  /**
   * Marca notificaciones como leídas.
   * Si se proporcionan ids, solo marca esas; si no, marca todas.
   *
   * @param userId — Id del usuario autenticado.
   * @param ids — Ids de notificaciones específicas (opcional).
   */
  async execute(userId: string, ids?: string[]): Promise<void> {
    if (ids && ids.length > 0) {
      await this.notifRepo.markReadByIds(userId, ids);
    } else {
      await this.notifRepo.markAllReadByUserId(userId);
    }
  }
}

File: FrontEnd/lib/domain/usecases/notifications/mark_notifications_read_use_case.dart
/// @file mark_notifications_read_use_case.dart
/// @description Caso de uso para marcar todas las notificaciones del usuario como leídas.
/// @module Reminders
/// @layer Domain
library;

import '../../interfaces/usecases/notifications/i_mark_notifications_read_use_case.dart';
import '../../repositories/i_notification_repository.dart';

// ═══════════════════════════════════════════════════════════════════════════════
// MARK NOTIFICATIONS READ USE CASE
// ═══════════════════════════════════════════════════════════════════════════════

/// Delega en [INotificationRepository.markAllRead].
///
/// [implements] IMarkNotificationsReadUseCase
/// [dependencies] INotificationRepository
class MarkNotificationsReadUseCase implements IMarkNotificationsReadUseCase {
  final INotificationRepository _repository;

  const MarkNotificationsReadUseCase({required INotificationRepository repository})
      : _repository = repository;

  @override
  Future<void> execute({List<String>? ids}) => _repository.markAllRead(ids: ids);
}
